### PR TITLE
enable codeql workflow

### PR DESCRIPTION
**Pull request by nautaa**
### Task Description
enable codeql workflow

### CODE REVIEW

File: .github/workflows/codeql.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    - name: Build init
+      run: |
+        bash build.sh init
+
+    - name: Build project
+      shell: bash
+      run: |
+        rm -rf build_debug
+        bash build.sh debug
+        cd build_debug && make -j4 && cd -
```
Comment (hnwyllmm) [inline .github/workflows/codeql.yml:54]: Does the build action can use ccache?

**Comment by hnwyllmm** [comment]
CodeQL takes a long time. So I think it's better to do the action periodically.

**Comment by caifeizhi** [comment]
Thank you for your contribution. Can you explain the advantages of adding CodeQL workflow?

File: .github/workflows/codeql.yml (hunk 1 above)
Comment (nautaa) [reply .github/workflows/codeql.yml:54]: > Does the build action can use ccache?

no, CodeQL requires a complete build to take place in order to perform analysis. ref: https://docs.github.com/en/code-security/code-scanning/troubleshooting-code-scanning/no-source-code-seen-during-build

**Comment by nautaa** [comment]
> CodeQL takes a long time. So I think it's better to do the action periodically.

Ok, in my tests, a full run took about 4 hours.

**Comment by nautaa** [comment]
> Thank you for your contribution. Can you explain the advantages of adding CodeQL workflow?

CodeQL help us find potential vulnerabilities in code(such as potential type overflow). We can find the result of CodeQL in Security->Code Scanning of repo.

**Comment by caifeizhi** [comment]
> CodeQL help us find potential vulnerabilities in code(such as potential type overflow). We can find the result of CodeQL in Security->Code Scanning of repo.

I see; And i agree with @hnwyllmm suggestion, you can change it to  be executed periodically.

File: .github/workflows/codeql.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      with:
+        key: ${{ runner.os }}-build-${{ env.cache-name }}-el9.x86_64-${{ hashFiles('deps/init/oceanbase.el9.x86_64.deps') }}
+        path: deps/3rd
+
+    - name: Initialize CodeQL
+      uses: github/codeql-action/init@v2
+      with:
+        languages: cpp
+
+    - name: Build init
```
Comment (caifeizhi) [inline .github/workflows/codeql.yml:45]: we can use one-line to do this? `bash build.sh --init --make -j4`

File: .github/workflows/codeql.yml (hunk 1 above)
Comment (hnwyllmm) [reply .github/workflows/codeql.yml:54]: Sorry, can you explain the words 'a complete build`?
I didn't find some information about it.
And another question, what's the difference between build without cache and build with cache to CodeQL?

Comment (hnwyllmm) [reply .github/workflows/codeql.yml:54]: Here is an interesting information:
https://codeql.github.com/docs/codeql-overview/supported-languages-and-frameworks/.
I think that CodeQL would use compiler to extract some information about the source code, so it requires `a complete build`.

**Review by hnwyllmm — APPROVED** [review]

File: .github/workflows/codeql.yml (hunk 2 above)
Comment (nautaa) [reply .github/workflows/codeql.yml:45]: > we can use one-line to do this? `bash build.sh --init --make -j4`

it's ok, actually these codes of build are reference the `complie.yml`.

File: .github/workflows/codeql.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    - name: Initialize CodeQL
+      uses: github/codeql-action/init@v2
+      with:
+        languages: cpp
+
+    - name: Build project
+      shell: bash
+      run: |
+        rm -rf build_debug
+        bash build.sh debug --init --make -j3
```
Comment (nautaa) [inline .github/workflows/codeql.yml:47]: The build is flaky with `make -j4`. So change this to `make -j3` in this action.
> The hosted runner: GitHub Actions 3 lost communication with the server. Anything in your workflow that terminates the runner process, starves it for CPU/Memory, or blocks its network access can cause this error.

**Pull request merged**